src/mgr/OSDPerfMetricCollector.cc

```diff
@@ -0,0 +1,147 @@
+// -*- mode:C++; tab-width:8; c-basic-offset:2; indent-tabs-mode:t -*-
+// vim: ts=8 sw=2 smarttab
+
+#include "common/Clock.h"
+#include "common/debug.h"
+#include "common/errno.h"
+
+#include "OSDPerfMetricCollector.h"
+
+#define dout_context g_ceph_context
+#define dout_subsys ceph_subsys_mgr
+#undef dout_prefix
+#define dout_prefix *_dout << "mgr.osd_perf_metric_collector " << __func__ << " "
+
+OSDPerfMetricCollector::OSDPerfMetricCollector(Listener &listener)
+  : listener(listener), lock("OSDPerfMetricCollector::lock"),
+    next_query_id(ceph_clock_now().to_nsec()) {
+}
+
+
+std::list<OSDPerfMetricQuery> OSDPerfMetricCollector::get_queries() {
+  Mutex::Locker locker(lock);
+
+  std::list<OSDPerfMetricQuery> query_list;
+  for (auto &it : queries) {
+    query_list.push_back(it.first);
+  }
+
+  return query_list;
+}
+
+uint64_t OSDPerfMetricCollector::add_query(const OSDPerfMetricQuery& query) {
+  uint64_t query_id;
+  bool notify = false;
+
+  {
+    Mutex::Locker locker(lock);
+
+    query_id = next_query_id++;
```

Comment:
Let's make this first compare with existing queries, and return the existing ID if found.  The CLI operations need to be idempotent, because the client logic resends commands on failovers etc.

Reply:
Then what would we do  on a client (one of the severals with the same query) unregisters (removes) its query? I think we can avoid `query_id` by using a refcount for every query registered. But again the issue with duplicates remains. Eventually it would expire but it does not look nice. Any ideas?

Reply:
The simplest thing is to let the clients specify the ID, rather than generating it on creation.  When you generate the ID on creation, you always have this problem of what to do when a client gets interrupted and retries.

Reply:
ok, I will do this way. Thanks.